File: HashGame.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
using namespace std;

class Game {
    char x = 'x';
    char o = 'o';
    char player;
    char computer;
    
public:
    Game() {
        char pos[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8'};
        bool running = true;
        
        cout << "Welcome to the game" << endl;
        
        // Who's playing
        int playerChoice;
        while (true) {
            cout << "\nHow would you like to play? \nPlayer x Computer -- 1\nPlayer x Player -- 2\n";
            try {
                cin >> playerChoice;
                
                if (playerChoice == 1 || playerChoice == 2) {
                    break;
                } else {
                    cout << "Please choose a valid option\n";
                }
            } catch(...) {
                cout << "Please insert a number\n";
            }
        }
        
        // Choose x or o
        while (true) {
            cout << "\nWould you like to play as X or O?\n x -- 1 \n o -- 0\n";

            if (playerChoice == 2) {
                cout << "Player 2 will receive the remaining option\n";
            }
            
            int pp;
            cin >> pp;
            cout << endl;
            
            if (pp == 1) {
                player = x;
                computer = o;
                break;
            } else if (pp == 0) {
                player = o;
                computer = x;
                break;
            } else {
                cout << "\nPlease insert a valid option\n";
            }
        }
        
        // Game loop
        while (running) {
            cout << pos[0] << "|" << pos[1] << "|" << pos[2] << endl;
            cout << "_____" << endl;
            cout << pos[3] << "|" << pos[4] << "|" << pos[5] << endl;
            cout << "_____" << endl;
            cout << pos[6] << "|" << pos[7] << "|" << pos[8] << endl;
            
            cout << "\nr - restart\nq - quit\nChoose the position you wish to insert: ";
            string getPos;
            cin >> getPos;
            
            if (getPos == "q") {
                running = false;
            } else if (getPos == "r") {
                char resetPos[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8'};
                for (int i = 0; i < 9; i++) {
                    pos[i] = resetPos[i];
                }
            } else {
                validAnswer(getPos, pos, playerChoice);
                
                // check win condition
                if ((pos[0] == pos[1] && pos[1] == pos[2]) ||
                    (pos[0] == pos[4] && pos[4] == pos[8]) ||
                    (pos[0] == pos[3] && pos[3] == pos[6]) ||
                    (pos[3] == pos[4] && pos[4] == pos[5]) ||
                    (pos[6] == pos[7] && pos[7] == pos[8]) ||
                    (pos[2] == pos[4] && pos[4] == pos[6]) ||
                    (pos[1] == pos[4] && pos[4] == pos[7]) ||
                    (pos[2] == pos[5] && pos[5] == pos[8])) {
                    cout << pos[0] << "|" << pos[1] << "|" << pos[2] << endl;
                    cout << "_____" << endl;
                    cout << pos[3] << "|" << pos[4] << "|" << pos[5] << endl;
                    cout << "_____" << endl;
                    cout << pos[6] << "|" << pos[7] << "|" << pos[8] << endl;
                    cout << "Game ended" << endl;
                    break;
                }
            }
        }
    }
    
    void validAnswer(string getPos, char* pos, int playerChoice) {
        try {
            int position = stoi(getPos);
            
            if (pos[position] == computer || pos[position] == player) {
                cout << "\nPlease choose a position that has not been occupied!\n";
            } else {
                pos[position] = player;
                
                if (playerChoice == 2) {
                    while (true) {
                        cout << pos[0] << "|" << pos[1] << "|" << pos[2] << endl;
                        cout << "_____" << endl;
                        cout << pos[3] << "|" << pos[4] << "|" << pos[5] << endl;
                        cout << "_____" << endl;
                        cout << pos[6] << "|" << pos[7] << "|" << pos[8] << endl;
                        
                        cout << "\nPlease choose a place: ";
                        int playerTwo;
                        cin >> playerTwo;
                        
                        if (pos[playerTwo] == computer || pos[playerTwo] == player) {
                            cout << "\nPlease choose a position that has not been occupied!\n";
                        } else {
                            pos[playerTwo] = computer;
                            break;
                        }
                    }
                } else {
                    while (true) {
                        int inList = rand() % 9;
                        
                        if (pos[inList] != 'x' && pos[inList] != 'o') {
                            pos[inList] = computer;
                            break;
                        } else {
                            continue;
                        }
                    }
                }
            }
        } catch (const invalid_argument& e) {
            cout << "\nPlease insert a valid position\n";
        }
    }
};

int main() {
    srand(time(0));
    Game game;
    return 0;
}

File: HashGame.py
from random import randint

class Game():

    x = "x"
    o = "o"
    player = False
    computer = False
    
    def __init__(self):
        
        pos = [1, 2, 3, 4, 5, 6, 7, 8, 9]
        running = True
    
        print("Welcome to the game \n\n")

        # Who's playing
        while True:
            print("\nHow would you like to play? \nPlayer x Computer -- 1\nPlayer x Player -- 2\n")
            try:

                playerChoice = int(input())
                
                if playerChoice == 1:
                    break
                elif playerChoice == 2:
                    break
                else:
                    print("Please choose a valid option")

            except:
                print("Please insert a number")
        
        # Choose x or o
        while True:
            print("\nWould you like to play as X or O?\n x -- 1 \n o -- 0\n")

            if playerChoice == 2:
                print("Player 2 will recieve the remaining option\n")

            try:
                pp = int(input())


                print("\n")
    
                if  pp == 1:
                    self.player = self.x
                    self.computer = self.o
                    break

                elif pp == 0:
                    self.player = self.o
                    self.computer = self.x
                    break
                else:
                    print("\nPlease insert a valid option\n")
            except:
                print("\nPlease choose a valid option\n")


        # Main Game
        while running:

            print(f"\n{pos[0]}|{pos[1]}|{pos[2]} \n_____\n{pos[3]}|{pos[4]}|{pos[5]}\n______\n{pos[6]}|{pos[7]}|{pos[8]}\n")


            print("r - restart\nq - quit\n\nChoose the position you wish to insert:\n")


            getPos = input()

            if getPos == "q":
                running = False

            elif getPos == "r":
                pos = [1, 2,  3, 4, 5, 6, 7, 8, 9] 

            else:
                self.validAnswer(getPos, pos, playerChoice)

                # check win condition
                if pos[0] == pos[1] == pos[2] or pos[0] == pos[4] == pos[8] or pos[0] == pos[3] == pos[6] or pos[3] == pos[4] == pos [5] or pos[6] == pos[7] == pos[8] or pos[2] == pos[4] == pos[6] or pos[1] == pos[4] == pos[7] or pos[2] == pos[5] == pos[8]:
            
                    print(f"\n{pos[0]}|{pos[1]}|{pos[2]} \n_____\n{pos[3]}|{pos[4]}|{pos[5]}\n______\n{pos[6]}|{pos[7]}|{pos[8]}\n")
                    print("Game ended")
                    break

    

    def validAnswer(self, getPos, pos, playerChoice):


        if playerChoice == 1:
    
            try:
    
                getPos = int(getPos)
                getPos = getPos - 1

                if pos[getPos]== self.computer or pos[getPos] == self.player:
                    print("\nPlease choose a position that has not been occupied!\n")

                else:
    
                    pos[getPos] = self.player

                    while True:
                
                        inList = randint(0, 8)

                        if inList in pos:
                            pos[inList] = self.computer
                            break
        
                        else:
                            continue

            except:
                
                    print("\nPlease insert a valuable postition\n")
    




        elif playerChoice == 2:
        
            try:
                getPos = int(getPos)
                getPos -= 1

                if pos[getPos] == self.computer or pos[getPos] == self.player:
                    print("\nPlease choose a position that has not been occupied!\n")

                else:
                    pos[getPos] = self.player
    
    

                    while True:
                    
                        try:
                            print(f"\n{pos[0]}|{pos[1]}|{pos[2]} \n_____\n{pos[3]}|{pos[4]}|{pos[5]}\n______\n{pos[6]}|{pos[7]}|{pos[8]}\n")
                        
                            print("\nPlease choose a place:\n")
                            playerTwo = int(input())
                            playerTwo -= 1

                            if pos[playerTwo] == self.computer or pos[playerTwo] == self.player:
                        
                                print("\nPlease choose a position that has not been occupied!\n")

                            else:
                                pos[playerTwo] = self.computer
                                break
                        except:
                            print("Please insert a valid number")


            except:
                print("\nPlease insert a valuable postition\n")

    

Game()
